examples/algorithm_progression.py

```diff
@@ -0,0 +1,172 @@
+import argparse
+import concurrent.futures
+import inspect
+import itertools
+import os
+from pathlib import Path
+
+import algorithms
+import numpy as np
+import yaml
+
+import large_image
+
+VARIABLE_LAYERS = ['z', 'c', 't']
+
+
+argparser = argparse.ArgumentParser(
+    prog='Algorithm Progression',
+    description='Apply an algorithm to an input image '
+                'for every parameter set in a given parameter space',
+)
+argparser.add_argument(
+    'algorithm_code',
+    choices=algorithms.ALGORITHM_CODES.keys(),
+    help='Code to specify which of the available algorithms should be used',
+)
+argparser.add_argument('input_filename', help='Path to an image to use as input')
+argparser.add_argument(
+    '-o',
+    '--output_dir',
+    required=False,
+    help='Name for a new directory in this location, wherein result images will be stored',
+)
+argparser.add_argument(
+    '-w',
+    '--num_workers',
+    required=False,
+    default=4,
+    type=int,
+    help='Number of workers to use for processing alogrithm iterations',
+)
+argparser.add_argument(
+    '-p',
+    '--param',
+    action='append',
+    required=False,
+    nargs='*',
+    help='A parameter to pass to the algorithm; '
+         'instead of using the default value, '
+         'Pass every item in the number space, '
+         'specified as `--param=param_name,start,end,num_items[,open]`',
+)
+
+
+def apply_algorithm(algorithm, input_filename, output_dir, params, unique_params):
+    source = large_image.open(input_filename)
+    filename = f'{algorithm.__name__}_{"_".join(["{:.2f}".format(v) for v in unique_params])}.tiff'
+    filepath = Path(output_dir, filename)
+    desc = {'path': filename}
+    for i, v in enumerate(unique_params):
+        desc[VARIABLE_LAYERS[i]] = float(v)
```

Comment:
We actually want the values for `z`, `t`, etc. to be 0-based integers, and not the floating point parameter values.

Reply:
Can we add arbitrary keys to the description that contain the true (float) values of the params, then? z, t, and c would be unique indexes, but we need some way to attach the true values to each result image.

Reply:
Yes -- You can arbitrary keys at the frame level.  There isn't any standard on different WSI formats on how to store actual values.  Maybe add a `parameters` dictionary with `hue_value: 0.123`, etc. in the is example and `parameters_axes` with `hue_value: z` (or `z: hue_value`?).

Reply:
Those values will presently be mostly ignored by the reader (ideally we'd eventually try to impose some consistency between tile sources on this process -- OME Tiff has some examples, but no regularity).

Reply:
For example, could the content of `results.yml` look like this?
```
- path: positive_pixel_count_1.00000_0.05000.tiff
  hue_width: 1.0
  z: 1
  hue_value: 0.05
  c: 0
- path: positive_pixel_count_0.00000_0.15000.tiff
  hue_value: 0.0
  z: 0
  hue_width: 0.15000000000000002
  c: 1
- path: positive_pixel_count_1.00000_0.15000.tiff
  hue_width: 0.15000000000000002
  z: 1
  c: 1
- path: positive_pixel_count_0.00000_0.05000.tiff
  hue_value: 0.05
  z: 0
  c: 0
```

Reply:
Yes, but I think it'd be better to nest the algorithm values under a subkey and indicate which axes have been used:
```
- path: positive_pixel_count_1_1.tiff
  parameters:
    values:
      hue_value: 1.0
      hue_width: 0.15000000000000002
      saturation_minimum: 0.05
      intensity_upper_limit: 0.95
      intensity_lower_limit: 0.05
      intensity_weak_threshold: 0.65
      intensity_strong_threshold: 0.35
    axes:
      z: hue_value
      c: hue_width
  z: 1
  c: 1
```
Part of my reasoning is that we have expanded what can be specified in the multi-source source in the past, so limiting the frame level keys might be better.

And, if we are sweeping many parameters, we can use custom axis names (add a top-level key `axes` with a list of axes names, then you can use those axes just like `z`, `t`, `xy`, and `c`.  You shouldn't use `x`, `y`, `s`, or `f`.

Reply:
Nevermind, I remembered that large_image does not accept arbitrary keys in each source, so I just put all that information in the description. `results.yml` now looks like this:

```
name: Positive Pixel Count iterative results
description: Positive Pixel Count algorithm performed on /home/anne/Documents/testing_grounds/large_image/data/sample_jp2k_33003_TCGA-CV-7242-11A-01-TS1.1838afb1-9eee-4a70-9ae3-50e3ab45e242.svs,
  where z represents change in hue_value as an index of [0. 1.],c represents change
  in hue_width as an index of [0.05 0.15],saturation_minimum remains unchanged per
  run with a value of 0.05,intensity_upper_limit remains unchanged per run with a
  value of 0.95,intensity_lower_limit remains unchanged per run with a value of 0.05,intensity_weak_threshold
  remains unchanged per run with a value of 0.65,intensity_strong_threshold remains
  unchanged per run with a value of 0.35
sources:
- path: /home/anne/Documents/testing_grounds/large_image/data/sample_jp2k_33003_TCGA-CV-7242-11A-01-TS1.1838afb1-9eee-4a70-9ae3-50e3ab45e242.svs
- path: positive_pixel_count_1_1.tiff
  z: 1
  c: 1
- path: positive_pixel_count_0_1.tiff
  z: 0
  c: 1
- path: positive_pixel_count_1_0.tiff
  z: 1
  c: 0
- path: positive_pixel_count_0_0.tiff
  z: 0
  c: 0

```

This is successful with the checking command you provided below, and the output of that command is `23021 23162 4`.